File: backend/utils/validators.py
# utils/validators.py — Input validation helpers
from config import ActiveConfig


def validate_register_payload(data: dict) -> str | None:
    """
    Validates registration data.
    Returns an error string if invalid, or None if OK.
    """
    username = (data.get("username") or "").strip()
    password = (data.get("password") or "")
    role     = (data.get("role") or "").strip().lower()
    branch   = (data.get("branch") or "").strip()

    if not username:
        return "Username is required."
    if len(username) < 3:
        return "Username must be at least 3 characters."
    if len(username) > 80:
        return "Username must be at most 80 characters."
    if " " in username:
        return "Username must not contain spaces."
    if not password:
        return "Password is required."
    if len(password) < 6:
        return "Password must be at least 6 characters."
    if role not in ActiveConfig.ALLOWED_ROLES:
        return f"Role must be one of: {', '.join(ActiveConfig.ALLOWED_ROLES)}."
    if branch not in ActiveConfig.ALLOWED_BRANCHES:
        return f"Branch must be one of: {', '.join(ActiveConfig.ALLOWED_BRANCHES)}."

    return None

File: frontend/src/utils/validators.js
/**
 * src/utils/validators.js
 * Reusable field-level validators for useFormValidation.
 */

export const required = (label = 'This field') =>
  (v) => (!v || !String(v).trim()) ? `${label} is required` : ''

export const minLength = (min, label = 'Value') =>
  (v) => v && v.length < min ? `${label} must be at least ${min} characters` : ''

export const maxLength = (max, label = 'Value') =>
  (v) => v && v.length > max ? `${label} must be at most ${max} characters` : ''

export const noSpaces = (label = 'Value') =>
  (v) => v && /\s/.test(v) ? `${label} must not contain spaces` : ''

export const range = (min, max, label = 'Value') =>
  (v) => {
    const n = parseFloat(v)
    if (isNaN(n)) return `${label} must be a number`
    if (n < min || n > max) return `${label} must be between ${min} and ${max}`
    return ''
  }

/** Chain multiple validators, return first error */
export const compose = (...validators) =>
  (v, all) => validators.reduce((err, fn) => err || fn(v, all), '')

/** Auth field validators */
export const authValidators = {
  username: compose(
    required('Username'),
    minLength(3, 'Username'),
    noSpaces('Username'),
  ),
  password: compose(
    required('Password'),
    minLength(6, 'Password'),
  ),
  role:   required('Role'),
  branch: required('Branch'),
}
